### ISSUE TITLE
Bluesky and AP: preserve whitespace?

### ISSUE THREAD

**Issue by snarfed** [snarfed/bridgy-fed#990]
Bluesky post content is plain text, so newlines and whitespace are meaningful, but AS2 content is generally HTML, so whitespace isn't meaningful. Should we convert Bluesky newlines to `<br>`s or similar?

We probably need to do the reverse in the other direction too.

cc @bnewbold

**Comment by bnewbold** [comment]
Yup! That sounds right to me. I think a single `<br>` is probably best, instead of a `<p>` or other options.

**Comment by mackuba** [comment]
Yeah…
<img width="550" alt="Screen Shot 2024-05-09 at 00 29 41" src="https://github.com/snarfed/bridgy-fed/assets/28465/4d6731e2-5f56-4080-b217-6db65b74efe7">
<img width="608" alt="Screen Shot 2024-05-09 at 00 29 32" src="https://github.com/snarfed/bridgy-fed/assets/28465/dd76ee80-8895-4b87-8581-8703be7fb64e">

**Comment by snarfed** [comment]
To reuse: https://github.com/snarfed/granary/blob/ed200452d6635b67fe649e124818d74701de44de/granary/microformats2.py#L1014-L1022

**Comment by snarfed** [comment]
Fixed! Example: https://bsky.app/profile/snarfed.bsky.social/post/3kswyx4strz2h => https://indieweb.social/@snarfed.bsky.social@bsky.brid.gy/112475072471845294 (requires login)

<img width="308" alt="image" src="https://github.com/snarfed/bridgy-fed/assets/778068/fe245438-6276-40a1-9877-29852e3d7f85">
<img width="231" alt="image" src="https://github.com/snarfed/bridgy-fed/assets/778068/77e72e61-8306-4bbd-ab0e-be374717c6fc">


Took a different approach here and abused an implementation detail of fediverse servers: that they apply a `white-space: pre-wrap` style to `<p>`s inside content. Awkward overreach, but useful since `white-space: pre-wrap` preserves more whitespace, eg leading spaces on a line, multiple spaces, etc, than just converting newlines to `<br>`s. 😎

**Comment by Tamschi** [comment]
This broke last-line hashtag de-emphasis in Mastodon.

Native (expected):
![small hashtags below the attachment, only three shown by default](https://github.com/snarfed/bridgy-fed/assets/715360/4b8b46df-6505-4bf0-b1e1-7ba9b7ca1ce4)

Bridged:
![large hashtags as part of the main text, all visible](https://github.com/snarfed/bridgy-fed/assets/715360/618cca28-5bc1-49ca-b063-c7c5040bc4a1)

Nothing critical is broken, but it would be nice to have the native presentation here since having all hashtags in the body like that is slightly annoying to come across.

**Comment by snarfed** [comment]
Oh interesting! Thanks for reporting, good point. Looks like this is https://bsky.app/profile/tamschi.bsky.social/post/3ksxa6sa7l22p => https://mastodon.gamedev.place/@tamschi.bsky.social@bsky.brid.gy/112475582044797265 . I'll file a new issue.

**Comment by TomCasavant** [comment]
This might be a bug in the apps themselves (because it looked fine in the web app). But it seems to remove the newlines in the official mastodon (android) app, moshidon (fork of mastodon) and fedilab. 

![Screenshot_20241017_160320_Tom-Moshidon Debug.jpg](https://github.com/user-attachments/assets/8f380173-60ee-430c-838d-be71adb6233d)

![Screenshot_20241017_160331_Bluesky.jpg](https://github.com/user-attachments/assets/bdd51795-b3e6-4263-90ef-5f0f97de662a)

https://fed.brid.gy/r/https://bsky.app/profile/did:plc:oio4hkxaop4ao4wz2pp3f4cr/post/3l6py2ij2kv2u

**Comment by snarfed** [comment]
Argh, yeah, I guess this is expected due to the way I fixed this, https://github.com/snarfed/bridgy-fed/issues/990#issuecomment-2121071809 :

> Took a different approach here and abused an implementation detail of fediverse servers: that they apply a white-space: pre-wrap style to <p>s inside content. Awkward overreach, but useful since white-space: pre-wrap preserves more whitespace, eg leading spaces on a line, multiple spaces, etc, than just converting newlines to <br>s. 😎

Re-opening, guess I need to do this right. 😐

**Comment by aciccarello** [comment]
Another counter-example, but starting from web. I recently noticed HTML-ignored new lines in posts published from my website are appearing in Mastodon but are rendered as expected in Bluesky (web). This causes my mastodon posts to have awkward line breaks.

### My site
[Original link](https://www.ciccarello.me/posts/2024/11/26/jam-bleachers/)
<img width="500" alt="Screenshot of a post with a paragraph of text all together." src="https://github.com/user-attachments/assets/25b99b4e-6718-495a-96dd-8839c5343e74">


### Mastodon (problematic)
[Mastodon link on indieweb.social](https://indieweb.social/@anthony@ciccarello.me/113549823575288443) (may get redirected if not logged in)
<img width="590" alt="Screenshot of the above post but in Mastodon, showing each sentence on its own line leading to awkward line breaks." src="https://github.com/user-attachments/assets/dee6927c-d88b-4318-b18e-c3ffa5c32826">

### Bluesky
[bsky link](https://bsky.app/profile/ciccarello.me.web.brid.gy/post/3lbuc3gssmb52)
<img width="583" alt="Screenshot of the same post but in Bluesky, showing the original paragraph format." src="https://github.com/user-attachments/assets/93175aab-c33b-4534-a24e-6812e7b01789">

**Comment by snarfed** [comment]
Converting plain text to HTML is pretty straightforward, esp if we focus just on whitespace, and if we don't try to handle tabs or multiple spaces inside content, eg `Hello     world`. I _think_ we'd just HTML-escape the plain text and then replace `\n` => '<br>' and leading spaces to `&nbsp;'.

...except, AS1 `tags` are facet-based, ie `startIndex`/`length`, so we have to call `as2.link_tags` first...which inserts HTML tags that we _don't_ want to escape. Ugh.

**Comment by snarfed** [comment]
OK, I think I"ve done this right now. Should be out in prod in 5m. 🤞🤞🤞

**Comment by snarfed** [comment]
Looking good! eg https://indieweb.social/@snarfed.bsky.social@bsky.brid.gy/116762432084495576

<img width="354" height="191" alt="Image" src="https://github.com/user-attachments/assets/68f3c45f-7196-4318-a5d0-5322c5022105" />
